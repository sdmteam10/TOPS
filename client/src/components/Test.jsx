import React, { Component } from 'react'
import Alert from "./Alert";




class Test extends Component {


  state = {
    status: false,
    technique: "",
    tempcodingTests: [],

    //*Q8 user input
    userInputTech: "",
    checked: false,
    codingTests: [],
    inputTechValues: []

  }

  successInfo =  () =>{
    Alert.open({
      alertTip:"Successful submitted. You may make changes and submit again.",
      closeAlert:function(){
          console.log("Alert is closed...");
      }
    });
  } 

  //This method is called by Q8 other checkbox.
  async setOtherTechValue(e){
    await this.setState({
      userInputTech: ""
    })
    if (e) {
      await this.setState ({checked: true})
    }else{
      await this.setState ({checked: false})
    }
 }
 
 async checkLastValue(event){
    this.state.inputTechValues.push(event.target.value)
    await this.setState({userInputTech: this.state.inputTechValues.pop()})

 }

 //Click submit button to submit the form
 submitForm = event => {

  event.preventDefault();
  //Push user input other values seperately when click on submit button
  if (this.state.userInputTech){
    this.state.codingTests.push(this.state.userInputTech)
  }
  console.log(this.state.codingTests)

  //Remove submitted user input other value
  let index = this.state.codingTests.findIndex(e => e === this.state.userInputTech)
  if (index >= 0){
    this.state.codingTests.splice(index,1)}

 }
  
  //For Q8 multi-selection checkbox except the customised input value
  async setTechniques(event) {
    // console.log("This is event", event)
    await this.setState({
      status: event.target.checked
      
     })
     console.log(this.state.status)
    
    await this.setState({technique: event.target.value})
  
    if (this.state.status === true){
        this.state.tempcodingTests.push(this.state.technique)
    } else {
      this.state.tempcodingTests.splice(this.state.tempcodingTests.findIndex(e => e === this.state.technique),1)
    }

    await this.setState({
      codingTests: (this.removeDuplicate(this.state.tempcodingTests))
    })

  }

  removeDuplicate(originalArray) {
    let unique = new Set(originalArray)
    return [...unique]
  }



  render() {
    return (
      <> 
        <div className='container-fluid col-sm-6'>
          <form className="mx-auto mt-md-5 font-monospace" onSubmit={this.submitForm}>
            {/* Q8*/}
            <div className="mb-3">
              <div onChange={this.setTechniques.bind(this)}>

                <div className="form-check">
                  <input className="form-check-input" type="checkbox" value="Infrastructure as code" id="flexCheckDefault08"></input>
                  <label className="form-check-label" htmlFor="flexCheckDefault08">
                    i. Infrastructure as code
                  </label>
                </div>

                <div className="form-check">
                  <input className="form-check-input" type="checkbox" value="Static code analysis" id="flexCheckDefault09"></input>
                  <label className="form-check-label" htmlFor="flexCheckDefault09">
                    j. Static code analysis
                  </label>
                </div>
              </div>


              <div className="input-group mb-3" onChange={this.checkLastValue.bind(this)}>
                <div className="input-group-text">
                  <input className="form-check-input mt-0" type="checkbox" value="Other" aria-label="Checkbox for following text input" checked={this.state.checked || false} disabled />
                </div>
                <input type="text" className="form-control" placeholder="Please specify.." aria-label="Text input with checkbox" onChange={event => this.setOtherTechValue(event.target.value)}></input>
              </div>
              
            </div> 
            {/* Submit */}
            <div className="mb-3">
              <button className="btn btn-primary" type="submit" onClick={this.successInfo}>Submit form</button>
            </div>    

          </form>
        </div>
      </>
    )
  }
}

export default Test            