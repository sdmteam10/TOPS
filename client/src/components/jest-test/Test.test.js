// Test.test.js
import React from 'react'
import Test from '../Test'
import renderer from 'react-test-renderer';
import { fireEvent, render } from "@testing-library/react"

//it.only() "only test this <it>."
//it.skip() "not test this <it>."
// done() under expect(){} -  should be only used for async functions

  describe('Test <Test /> component', function() {
    it('Should render', function() {  
      //Render component should be successed.
      const component = renderer.create(
        <Test />
      );
        let testSnap = component.toJSON();
        expect(testSnap).toMatchSnapshot(); 
    })
  
  //Function should capture onChange changes into event.
    it('capture changes', function() {   //it.only() "only test this."

      function setTechniques(event) {
        expect(event.target.value).toEqual("whamo")
        done()
      }
      const { getByPlaceholderText }  = render(
        <div onChange={setTechniques}>
          <input placeholder="test input"/>
        </div>
      )
      fireEvent.change(getByPlaceholderText("test input"));

    })

    //Function should capture onSubmit into event.
    it('capture submit', function() {   

      function submitForm(event) {
        expect(event.target.value).toEqual("")
        
      }
      const { getByPlaceholderText }  = render(
        <form onSubmit={submitForm}>
          <input placeholder="test submit"/>
        </form>
      )
      fireEvent.submit(getByPlaceholderText("test submit"));

    })
    
    //Test removeDuplicate functionality
    it('remove duplicate string array -> abca should output abc', function() {
      const tree = renderer.create(<Test />);
      expect(tree.getInstance().removeDuplicate(["a","b","c","a"])).toEqual(["a","b","c"])
      
    })

    it('remove duplicate object in array', function() {
      const tree = renderer.create(<Test />)
      expect(tree.getInstance().removeDuplicateOBJ([
      { _id: '614a64afad8de7c962ddd661', name: 'Peer Support', activityNumber: 1, __v: 0 },
      { _id: '614a64daad8de7c962ddd663', name: 'Team Leader Assistance', activityNumber: 2, __v: 0 },
      { _id: '614a64eaad8de7c962ddd665', name: 'Induction', activityNumber: 3, __v: 0 },
      { _id: '614a64f3ad8de7c962ddd667', name: 'Agile Course', activityNumber: 4, __v: 0 },
      { _id: '614a6502ad8de7c962ddd669', name: 'Review Plan', activityNumber: 5, __v: 0 },
      { _id: '614a64f3ad8de7c962ddd667', name: 'Agile Course', activityNumber: 4, __v: 0 },
      { _id: '614a64daad8de7c962ddd663', name: 'Team Leader Assistance', activityNumber: 2, __v: 0 },
      { _id: '614a6513ad8de7c962ddd66b', name: 'Checklists', activityNumber: 6, __v: 0 },
      { _id: '614a651ead8de7c962ddd66d', name: 'Other Teams', activityNumber: 7, __v: 0 },
      { _id: '614a6502ad8de7c962ddd669', name: 'Review Plan', activityNumber: 5, __v: 0 }])).toEqual([
          { _id: '614a64afad8de7c962ddd661', name: 'Peer Support', activityNumber: 1, __v: 0 },
          { _id: '614a64daad8de7c962ddd663', name: 'Team Leader Assistance', activityNumber: 2, __v: 0 },
          { _id: '614a64eaad8de7c962ddd665', name: 'Induction', activityNumber: 3, __v: 0 },
          { _id: '614a64f3ad8de7c962ddd667', name: 'Agile Course', activityNumber: 4, __v: 0 },
          { _id: '614a6502ad8de7c962ddd669', name: 'Review Plan', activityNumber: 5, __v: 0 },
          { _id: '614a6513ad8de7c962ddd66b', name: 'Checklists', activityNumber: 6, __v: 0 },
          { _id: '614a651ead8de7c962ddd66d', name: 'Other Teams', activityNumber: 7, __v: 0 }
        ])
        

    })


  })


