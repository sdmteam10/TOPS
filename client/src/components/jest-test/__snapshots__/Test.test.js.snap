// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <Test /> component Should render 1`] = `
<div
  className="container-fluid col-sm-6"
>
  <form
    className="mx-auto mt-md-5 font-monospace"
    onSubmit={[Function]}
  >
    <div
      className="mb-3"
    >
      <div
        onChange={[Function]}
      >
        <div
          className="form-check"
        >
          <input
            className="form-check-input"
            id="flexCheckDefault08"
            type="checkbox"
            value="Infrastructure as code"
          />
          <label
            className="form-check-label"
            htmlFor="flexCheckDefault08"
          >
            i. Infrastructure as code
          </label>
        </div>
        <div
          className="form-check"
        >
          <input
            className="form-check-input"
            id="flexCheckDefault09"
            type="checkbox"
            value="Static code analysis"
          />
          <label
            className="form-check-label"
            htmlFor="flexCheckDefault09"
          >
            j. Static code analysis
          </label>
        </div>
      </div>
      <div
        className="input-group mb-3"
        onChange={[Function]}
      >
        <div
          className="input-group-text"
        >
          <input
            aria-label="Checkbox for following text input"
            checked={false}
            className="form-check-input mt-0"
            disabled={true}
            type="checkbox"
            value="Other"
          />
        </div>
        <input
          aria-label="Text input with checkbox"
          className="form-control"
          onChange={[Function]}
          placeholder="Please specify.."
          type="text"
        />
      </div>
    </div>
    <div
      className="mb-3"
    >
      <button
        className="btn btn-primary"
        onClick={[Function]}
        type="submit"
      >
        Submit form
      </button>
    </div>
  </form>
</div>
`;
